junit5/src/main/java/cz/xtf/junit5/listeners/OpenshiftStateRecorder.java

```diff
@@ -0,0 +1,199 @@
+
+import static org.junit.platform.engine.TestExecutionResult.Status.FAILED;
+
+/**
+ * Listener for storing information from openshift in case a test fails.
+ * It stores pod logs, openshift routes, secret, events etc. From both work and build namespace.
+ * It implements two distinct interfaces TestExecutionListener and LifecycleMethodExecutionExceptionHandler
+ * both serving different purpose and need to be applied differently.
+ *
+ * TestExecutionListener
+ *    Is called whenever test ends. Stores openshift state if test failed.
+ *    To use this interface add this class as a listener
+ *
+ * LifecycleMethodExecutionExceptionHandler
+ *    This interface is executed when @BeforeAll or @BeforeEach method fails - throw an exception.
+ *    To use this interface a test class needs to be extended with this class using @ExtendWith
+ */
+@Slf4j
+public class OpenshiftStateRecorder implements TestExecutionListener, LifecycleMethodExecutionExceptionHandler {
+	private static final Path statusDir = Paths.get("log","status");
+
+	private final OpenShift openShift = OpenShifts.master();
+	private final OpenShift buildOpenShift;
+
+	private final String namespace;
+	private final String buildNamespace;
+
+	public OpenshiftStateRecorder(){
+		namespace = openShift.getNamespace();
+		buildNamespace = XTFConfig.get(BuildManagerConfig.BUILD_NAMESPACE);
+		buildOpenShift = OpenShifts.master(buildNamespace);
+	}
+
+	private String getTestDisplayName(TestIdentifier testIdentifier) {
+		String className = testIdentifier.getParentId().get()
+			.replaceAll(".*class:", "")
+			.replaceAll("].*", "");
+		return String.format("%s#%s", className, testIdentifier.getDisplayName());
+	}
+
+	// Log the openshift state if BeforeAll fail
+	@Override
+	public void handleBeforeAllMethodExecutionException(final ExtensionContext context, final Throwable ex) throws Throwable {
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if BeforeEach fail
+	@Override
+	public void handleBeforeEachMethodExecutionException(final ExtensionContext context, final Throwable ex) throws Throwable {
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if test fail
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		try {
+			if (FAILED.equals(testExecutionResult.getStatus()) && testIdentifier.isTest()) {
+				storeOpenshiftState(getTestDisplayName(testIdentifier));
+			}
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private void storeOpenshiftState(String testIdentifier) throws IOException {
```

Comment:
I can imagine it could be useful to have method like this available in XTF public API. So when troubelshooting some test I can let store the state in some important points. E.g. before test and after test or some important place in the middle.

Reply:
Yes it could be worth it. 
Any idea what is the right place to put it in? I think that Openshift class might be good, but not sure if it fits here.